torch/ao/quantization/experimental/linear.py

```diff
@@ -0,0 +1,168 @@
+import torch
+import numpy as np
+
+from torch.nn.quantized.modules.utils import WeightedQuantizedModule
+from torch.ao.quantization.experimental.observer import APoTObserver
+from torch.ao.quantization.experimental.quantizer import quantize_APoT
+
+class LinearAPoT(WeightedQuantizedModule):
+    r"""
+    A quantized linear module with quantized tensor as inputs and outputs
+    to support APoT quantization.
+    We adopt the same interface as `torch.nn.Linear`, see
+    https://pytorch.org/docs/stable/nn.html#torch.nn.Linear for documentation.
+
+    Similar to :class:`~torch.nn.Linear`, attributes will be randomly
+    initialized at module creation time and will be overwritten later
+
+    Attributes:
+        alpha: `alpha` qparam of output Quantized Tensor, type: Tensor
+        gamma: `gamma` qparam of output Quantized Tensor, type: Tensor
+        quantization_levels: `quantization_levels` qparam of output Quantized Tensor, type: Tensor
+        level_indices: `level_indices` qparam of output Quantized Tensor, type: Tensor
+        weight: APoT quantized tensor from weight2quantize
+        weight_transposed: transposed weight tensor, used in linear transformation calculation (y = x * A^T + b)
+    """
+
+    def __init__(self, weight2quantize: torch.Tensor):
+        assert weight2quantize.dim() == 2
+
+        super().__init__()
+
+        # hard code b, k to match uniform quantization: b=4, k=1
+        self.b = 8
+        self.k = 1
+        self.n = 8
+
+        observer = APoTObserver(b=self.b, k=self.k)
+
+        observer(weight2quantize)
+
+        self.alpha, self.gamma, self.quantization_levels, self.level_indices = observer.calculate_qparams(signed=False)
+
+        quantized = quantize_APoT(weight2quantize, self.alpha, self.gamma, self.quantization_levels, self.level_indices)
+        self.weight = quantized.data
+        self.weight_transposed = torch.transpose(self.weight, 0, 1)
+
+        self.weight_transposed = self.weight_transposed.reshape(weight2quantize.shape)
+        self.weight_transposed = self.weight_transposed.reshape(weight2quantize.shape[1], weight2quantize.shape[0])
+
+    def decompose_APoT(self, x):
+        r"""
+        Decompose APoT quantized terms into binary digits
+        Args:
+            x (Tensor): binary representation of APoT quantized tensor
+        """
+        # remove "0b" prefix from binary representation
+        x = x[2:]
+
+        # initialize list of blocks
+        blocks = []
+
+        while x:
+            blocks.append(x[0:self.k])
+            x = x[self.k:]
+
+        return blocks
+
+    def bitshift_mul(self, decomposed_weight, activation):
```

Comment:
you're still doing the matrix multiplication in this function. it doesn't address my previous comment. what you have here isn't just doing bitshift_mul. it's doing the entire matrix multiplication